### PR TITLE
[FLINK-17869][task][checkpointing] Fix race condition when calling ChannelStateWriter.abort

### PR DESCRIPTION
**Pull request by rkhachatryan**
## What is the purpose of the change

Before FLINK-17218 there were two problems:
1. Exceeding `ChannelStateWriterImpl.maxCheckpoints`
1. Race condition with out of order barriers (see be8fbcf506b8dd38e5425cf772a55f033f0962b0 of this PR)

FLINK-17218 (24ff415f1b76392f75dea7c3538558d24fcb7058) fixed the first one but introduced a new race condition when netty thread modifies a ChannelStateWriter map used by the task thread.

This PR reverts FLINK-17218 and addresses the issues above.

## Brief change log
1. Revert 24ff415f1b76392f75dea7c3538558d24fcb7058 `"[FLINK-17218][checkpointing] Ensuring that ChannelStateWriter aborts previous checkpoints before a new checkpoint`
1. Abort channel state write if checkpoint is subsumed - using future callback and task thread - fix race condition
1. Increase `ChannelStateWriterImpl.DEFAULT_MAX_CHECKPOINTS` - fix the original problem of `FLINK-17218`
1. Ignore out of order checkpoints in `SubtaskCheckpointCoordinator` - fix a more subtle race condition
1. Unignore `UnalignedCheckpointITCase`

## Verifying this change

Unignore `UnalignedCheckpointITCase`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java [hunk 1]
```diff
@@ -158,6 +159,16 @@ private ChannelStateWriter openChannelStateWriter() {
...
+		lastCheckpointId = Math.max(lastCheckpointId, checkpointId);
+		Iterator<Long> iterator = abortedCheckpointIds.iterator();
+		while (iterator.hasNext()) {
+			long next = iterator.next();
+			if (next < lastCheckpointId) {
+				iterator.remove();
+			} else {
+				break;
+			}
+		}
```
Comment (rkhachatryan) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:171]: This is just an optimization.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java [hunk 2]
```diff
@@ -281,6 +281,8 @@ public void notifyCheckpointAborted(long checkpointId, OperatorChain<?, ?> opera
 				}
 			}
 
+			channelStateWriter.abort(checkpointId, new CancellationException("checkpoint aborted via notification"), false);
```
Comment (rkhachatryan) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:284]: This is just an optimization (possible now with `abort` and `cleanup=false`).

Comment (rkhachatryan) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:284]: This is just an optimization (possible now with `abort` and `cleanup=false`).

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java [hunk 3]
```diff
@@ -102,11 +102,11 @@ public void start(long checkpointId, CheckpointOptions checkpointOptions) {
 		LOG.debug("{} starting checkpoint {} ({})", taskName, checkpointId, checkpointOptions);
 		ChannelStateWriteResult result = new ChannelStateWriteResult();
 		ChannelStateWriteResult put = results.computeIfAbsent(checkpointId, id -> {
-			Preconditions.checkState(results.size() < maxCheckpoints, "results.size() > maxCheckpoints", results.size(), maxCheckpoints);
+			Preconditions.checkState(results.size() < maxCheckpoints, String.format("can't start %d, results.size() > maxCheckpoints: %d, %d, %s", checkpointId, results.size(), maxCheckpoints, taskName));
```
Comment (AHeise) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java:104]: nit: usually we add task name to front, also it would be nice to use `%d > %d` for the comparison values.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java [hunk 4]
```diff
@@ -56,7 +56,7 @@
 public class ChannelStateWriterImpl implements ChannelStateWriter {
 
 	private static final Logger LOG = LoggerFactory.getLogger(ChannelStateWriterImpl.class);
-	private static final int DEFAULT_MAX_CHECKPOINTS = 5; // currently, only single in-flight checkpoint is supported
+	private static final int DEFAULT_MAX_CHECKPOINTS = 100; // includes max-concurrent-checkpoints + checkpoints to be aborted (scheduled via mailbox)
```
Comment (AHeise) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java:59]: I also thought about that in FLINK-17218, but ultimately couldn't find a reason why any arbitrary large value offers an advantage over the current value. It makes the failing `checkState` more unlikely, but didn't eliminate it.

I also thought about removing the `checkState`, but it helped me to find some issues previously.

TL;DR I'm not convinced that this is a proper fix.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java [hunk 5]
```diff
@@ -196,9 +207,14 @@ public void checkpointState(
 		// We generally try to emit the checkpoint barrier as soon as possible to not affect downstream
 		// checkpoint alignments
 
+		if (lastCheckpointId >= metadata.getCheckpointId()) {
+			LOG.warn("Out of order checkpoint barrier (aborted previously?): {} >= {}", lastCheckpointId, metadata.getCheckpointId());
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:211]: Should that be warning? I think on concept level yes.
However, in reality, in current UC, this case happens probably quite frequently (barrier overtakes by cancellation marker doesn't). I wouldn't like seeing the log polluted with WARN.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java [hunk 6]
```diff
@@ -141,15 +141,10 @@ boolean isDone() {
 	void abort(long checkpointId, Throwable cause);
 
 	/**
-	 * Must be called after {@link #start(long, CheckpointOptions)}.
+	 * Must be called after {@link #start(long, CheckpointOptions)} once.
 	 */
 	ChannelStateWriteResult getWriteResult(long checkpointId);
```
Comment (AHeise) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java:147]: Please rename `getWriteResult`. I think most developers would expect a getter to be idempotent.

How about `removeWriteResult`? That would be similar to how a Java map works. Or `getAndRemoveWriteResult` to be completely explicit.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java [hunk 7]
```diff
@@ -137,8 +137,9 @@ boolean isDone() {
 
 	/**
 	 * Aborts the checkpoint and fails pending result for this checkpoint.
+	 * @param cleanup true if {@link #getWriteResult(long)} is not supposed to be called afterwards.
 	 */
-	void abort(long checkpointId, Throwable cause);
+	void abort(long checkpointId, Throwable cause, boolean cleanup);
```
Comment (AHeise) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java:142]: I guess the intend of cleanup is to avoid throwing an exception if `getWriteResult(long)` is called nonetheless, right?
So in which cases can that happen? Do we ensure cleanup in these cases (e.g., is abort then called twice?)?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java (hunk 2 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:284]: I like this optimization.
What happens if some other thread still enqueues data to this checkpoint? Or the writes dismissed?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java (hunk 6 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java:147]: Good point, I'll rename it to `getAndRemoveWriteResult`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java (hunk 4 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java:59]: The map is already cleared during the regular checkpoint abortion by the task thread but it happens with some delay. I see only two reasons why its size could significantly exceed `max-concurrent-checkpoints`:
1. bug in abort procedure
2. task thread is stuck while netty thread continues to receive new barriers fast

Theoretically, 2nd case is unbounded, but in practice, I didn't observe issues with a value of 5 in `UnalignedITCase` after fixing other issues. I think such a scenario where task thread is not aborting writes would signify some problem and should be investigated.

Besides that, aborting async parts of previous checkpoints upon receiving a new barrier won't work with `max-concurrent-checkpoints` > 1.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java (hunk 7 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java:142]: > I guess the intend of cleanup is to avoid throwing an exception if getWriteResult(long) is called nonetheless, right?

Yes.

> So in which cases can that happen?

`abort` with `cleanup=false` can be called when:
1. receiving abort notification via RPC
2. `CheckpointBarrierUnaligner.allBarriersReceivedFuture` is cancelled

> Do we ensure cleanup in these cases (e.g., is abort then called twice?)?

Yes, for each checkpoint either "regular" `abort` or `getWriteResult` is eventually called.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java (hunk 2 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:284]: Yes, write requests will be ignored when dequeued by `ChannelStateWriteRequestExecutor`.

**Comment by rkhachatryan** [comment]
Thanks for the feedback @AHeise ,
I've updated the PR and answered your comments, please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java (hunk 4 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java:59]: Just thinking loud.

Previously, the issue was that sources received new RPC triggers while being stuck, which enqueued a ton of mails.
With a checkpointing interval of 100ms, you only need to be stuck for 10s until you enqueue 100 mails and hit the limit.

But I guess, the assumption is that now `notifyCheckpointAborted` is called reliably. And if doesn't, we need to fail probably anyways, since something is broken.

So I guess this solution is as good as it gets. I trade-off between quickly identifying bugs in abort and running potentially into some issues with ultra slow tasks and ultra fast checkpoint barriers.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java [hunk 8]
```diff
@@ -186,7 +186,7 @@ public void abortCheckpointOnBarrier(long checkpointId, Throwable cause, Operato
 
 		checkpointStorage.clearCacheFor(checkpointId);
 
-		channelStateWriter.abort(checkpointId, cause);
+		channelStateWriter.abort(checkpointId, cause, true);
```
Comment (zhijiangW) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:200]: I am not quite sure whether we still have race condition here. `abortCheckpointOnBarrier` might be called from `CheckpointBarrierUnaligner` after triggering checkpoint into mailbox. After aborting, we did not remove the checkpoint action from mailbox, so the checkpoint might still happen afterwards. So how to guarantee that `#getAndRemoveWriteResult` would never be called after aborting?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java [hunk 9]
```diff
@@ -56,7 +56,7 @@
 public class ChannelStateWriterImpl implements ChannelStateWriter {
 
 	private static final Logger LOG = LoggerFactory.getLogger(ChannelStateWriterImpl.class);
-	private static final int DEFAULT_MAX_CHECKPOINTS = 5; // currently, only single in-flight checkpoint is supported
+	private static final int DEFAULT_MAX_CHECKPOINTS = 1000; // includes max-concurrent-checkpoints + checkpoints to be aborted (scheduled via mailbox)
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java:59]: I guess the current way seems a temporary work-around solution, not an elegant way. The initial purpose for introducing this threshold is for logic validating and avoiding invalid checkpoints retained in writer forever. But if we consider the abort delay into this threshold, it seems somehow lose the initial meaning for the guard, and we are really not sure what is the proper value for this threshold.

The proper way might resolve the potential race condition in essence, but it might pay more efforts not feasible ATM. So I think we might leave another debt here in future.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java (hunk 8 above)
Comment (zhijiangW) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:200]: I got the answer when seeing the following commit "Ignore out of order checkpoints in SubtaskCheckpointCoordinator". It updates the `lastCheckpointId` to prevent the following aborted checkpoint execution.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java [hunk 10]
```diff
@@ -221,9 +232,14 @@ public void checkpointState(
 		// We generally try to emit the checkpoint barrier as soon as possible to not affect downstream
 		// checkpoint alignments
 
+		if (lastCheckpointId >= metadata.getCheckpointId()) {
```
Comment (zhijiangW) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:235]: For the below logic for checking abort request from RPC:

```
if (checkAndClearAbortedStatus(metadata.getCheckpointId())) {
			LOG.info("Checkpoint {} has been notified as aborted, would not trigger any checkpoint.", metadata.getCheckpointId());
			return;
}
```
Do we also need to call `channelStateWriter.abort` for RPC abort? If so, maybe we can integrate these two cases together for better handling the actions for aborting.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java (hunk 9 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java:59]: Me and @AHeise also had these concerns and discussed it [above](https://github.com/apache/flink/pull/12478#discussion_r436502262) (I increased the limit to 1K since then).

I'm not sure if there __is__ a proper solution because in the general case (max-concurrent-checkpoints > 1) we simply don't know which checkpoints to abort. Another issue is thread-safety, which is probably solvable but with some extra complexity which I think doesn't pay off here.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java (hunk 10 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:235]: Yes, I added it as a separate commit (it is just an optimization because we'll still get a "regular" abort call).

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java (hunk 8 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:200]: Right, there are several race conditions that are addressed in different commits.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java (hunk 10 above)
Comment (zhijiangW) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:235]: In addition, it is a bit tough to understand the semantic of different cases now. The previous `lastCheckpointId` only reflects the already happened checkpoint, so wee can rely on `lastCheckpointId >= metadata.getCheckpointId()` to discard all the following smaller checkpoints based on the checkpoint id incremental guarantee.

Now the `lastCheckpointId` also reflects the canceled id from `CheckpointBarrierHandler`, so if the handler notifies to abort checkpoint 10, should we also need to ignore checkpoint 9. If so, it is also different with the abort case from RPC. If RPC notifies to abort checkpoint 10, the checkpoint 9 can still execute afterwards since we store the aborted id from RPC in the `abortedCheckpointIds` set.

Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:235]: Yes, this is the right semantics and I agree it is more complex now. 
I described it in the comment for the aborted checkpoints field. 
However, I think this new complexity comes from the domain (or maybe design), and therefore can't be avoided in implementation.

Comment (zhijiangW) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:235]: Yes, the current complexity somehow comes from coupling abort from handler with abort from RPC together, since `abortedCheckpointIds` will compare with `lastCheckpointId(aborted id from handler)` in some processes.

But indeed I can not think of an easy solution right now and also have not thought of critical problems ATM. So let's further discuss it if encountering problems future.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java (hunk 9 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java:59]: Ok, let's keep it as now since no better solutions right now. And further improve it if possible future.

**Comment by zhijiangW** [comment]
Thanks for the fix @rkhachatryan and reviews @AHeise .

I did not go through all the details of this PR and only discussed some potential concerns.
Since @AHeise already approved it, I will merge directly.
